{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Package imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transcript/ENSEMBL imports\n",
    "conv_trans = pd.read_csv('jupyter/transcript-gene-conversion.txt', index_col = 2)\n",
    "conv_id = pd.read_csv('jupyter/ensembl-geneid-conversion.txt', sep='\\t', index_col = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert ENSEMBL IDs to genes\n",
    "# Create a list of all of the transcripts to convert\n",
    "transcripts = list(dat_test.index)\n",
    "transcripts = [x[:-2] for x in transcripts]\n",
    "transcripts_list = list(conv_trans.index)\n",
    "\n",
    "# Replace ENSEMBL IDs with gene names, establishing the new index for all datasets\n",
    "n = len(transcripts)\n",
    "new_index = ['NaN']*n\n",
    "for i in range(n):\n",
    "    if (transcripts[i] in transcripts_list):\n",
    "        new_index[i] = conv_trans.loc[transcripts[i]]['Gene stable ID']\n",
    "    if (i%10000 == 0): # For troubleshooting\n",
    "        print(i)\n",
    "\n",
    "# Export the new index \n",
    "pd.DataFrame(new_index).to_csv('jupyter/index_ensembl_gene.txt')\n",
    "\n",
    "# Convert the ensembl gene id to the common gene id\n",
    "index_in = pd.read_csv('jupyter/index_ensembl_gene.txt', index_col = 1)\n",
    "index_in = list(index_in.index)\n",
    "genes_list = list(conv_id.index)\n",
    "\n",
    "n = len(index_in)\n",
    "final_index = ['NaN']*n\n",
    "for i in range(n):\n",
    "    gene = str(index_in[i])\n",
    "    if (gene in genes_list):\n",
    "        final_index[i] = conv_id.loc[gene]['Approved symbol']\n",
    "    if (i%10000 == 0): # For troubleshooting\n",
    "        print(i)\n",
    "\n",
    "# Export the final index\n",
    "pd.DataFrame(final_index).to_csv('jupyter/index_geneid.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the gene index\n",
    "index_gene = pd.read_csv('jupyter/index_geneid.txt', index_col = 1)\n",
    "index_gene = list(index_gene.index)\n",
    "\n",
    "# Define a function that takes in a dataframe in the abundance.tsv format, and outputs a dataframe that\n",
    "# adds the gene ID, sums \"duplicates\", and keeps only tpm values\n",
    "def reformat(df_in):\n",
    "    df = df_in\n",
    "    df['ID'] = index_gene\n",
    "    df = df.dropna()\n",
    "    df = df.set_index('ID')\n",
    "    df = df.drop(['length','eff_length','est_counts'], axis = 1)\n",
    "    \n",
    "    # Fix duplicates\n",
    "    df['gene'] = df.index\n",
    "    df = df.groupby(['gene']).sum()\n",
    "    \n",
    "    # Return the results\n",
    "    return(df)\n",
    "\n",
    "def reformat_counts(df_in):\n",
    "    df = df_in\n",
    "    df['ID'] = index_gene\n",
    "    df = df.dropna()\n",
    "    df = df.set_index('ID')\n",
    "    df = df.drop(['length','eff_length','tpm'], axis = 1)\n",
    "    \n",
    "    # Fix duplicates\n",
    "    df['gene'] = df.index\n",
    "    df = df.groupby(['gene']).sum()\n",
    "    \n",
    "    # Return the results\n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output_tpm and output_ct have identical columns? True\n"
     ]
    }
   ],
   "source": [
    "# Define a function that takes a sample, i.e. DTB-184, as an input, reads in the file, outputs a dataframe\n",
    "# with 'DTB-184' as the TPM column name, and merges that dataframe with the dataframe produced from the \n",
    "# prior iteration. Create an output tpm and similar output counts file\n",
    "output_tpm = pd.DataFrame();\n",
    "output_ct = pd.DataFrame();\n",
    "samples = ['DTB-021','DTB-175','DTB-184','DTB-188','PR-009','PR-028','PR-038','PR-051','PR-056','PR-058','PR-063','PR-080'];\n",
    "for samp in samples:\n",
    "    # Read in the tsv file\n",
    "    str_in = 'result-' + samp + '/abundance.tsv'\n",
    "    dat_in = pd.read_csv(str_in, sep='\\t', index_col = 0)\n",
    "    \n",
    "    # Reformat to fix duplicates and include only tpm; rename by sample\n",
    "    dat_new = reformat(dat_in)\n",
    "    dat_new = dat_new.rename(columns = {'tpm':samp})\n",
    "\n",
    "    dat_ct = reformat_counts(dat_in)\n",
    "    dat_ct = dat_ct.rename(columns = {'est_counts':samp})\n",
    "\n",
    "    # Append to the output dataframes\n",
    "    output_tpm = pd.concat([output_tpm, dat_new], axis=1, sort=False)\n",
    "    output_ct = pd.concat([output_ct, dat_ct], axis=1, sort=False)\n",
    "\n",
    "# Add a row with the total counts to output_ct    \n",
    "output_ct['sum'] = list(output_ct.transpose().sum())\n",
    "\n",
    "# Make a list of genes where counts exceed an average of 100 (i.e. > 1200 total counts in 12 samples)\n",
    "high_genes = list(output_ct[output_ct['sum'] >= 1200].index)\n",
    "\n",
    "# Check that the columns are identical\n",
    "test = list(output_tpm.index)\n",
    "test2 = list(output_ct.index)\n",
    "res = (test == test2)\n",
    "print('output_tpm and output_ct have identical columns? ' + str(res))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the output_tpm matrix to only include results with sufficiently high counts\n",
    "tpm_final = output_tpm.loc[high_genes]\n",
    "\n",
    "# Make dataset for top vs bottom half, bone only\n",
    "bone = ['DTB-188', 'PR-063', 'PR-038', 'PR-058', 'DTB-184', 'DTB-021', 'PR-051', 'PR-080']\n",
    "tpm_bone = tpm_final[bone]\n",
    "tpm_bone.to_csv('jupyter/tpm_citrate_bone.csv')\n",
    "\n",
    "# Make the design matrix based on GaCitrate uptake, split at median\n",
    "des_bone = list(tpm_bone.transpose().index)\n",
    "des_citrate3 = pd.DataFrame([1,1,1,1,0,0,0,0], columns = ['ga_low'])\n",
    "des_citrate3['ga_high'] = 1-des_citrate3['ga_low']\n",
    "des_citrate3.index = des_bone\n",
    "des_citrate3.index.name = 'samples'\n",
    "des_citrate3.to_csv('jupyter/des_citrate_bone.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ga_low</th>\n",
       "      <th>ga_high</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samples</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DTB-188</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PR-009</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PR-028</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PR-056</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DTB-184</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DTB-021</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PR-051</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PR-080</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ga_low  ga_high\n",
       "samples                 \n",
       "DTB-188       1        0\n",
       "PR-009        1        0\n",
       "PR-028        1        0\n",
       "PR-056        1        0\n",
       "DTB-184       0        1\n",
       "DTB-021       0        1\n",
       "PR-051        0        1\n",
       "PR-080        0        1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "des_citrate3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
